lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java

```diff
@@ -126,14 +163,40 @@ public ParquetReader(
             checkArgument(blocks.size() == firstRows.size(), "elements of firstRowsOfBlocks must correspond to blocks");
         });
 
-        Map<ChunkKey, DiskRange> ranges = new HashMap<>();
+        this.columnIndexStore = columnIndexStore;
+        this.blockRowRanges = listWithNulls(this.blocks.size());
+        for (PrimitiveColumnIO column : columns) {
+            ColumnDescriptor columnDescriptor = column.getColumnDescriptor();
+            this.paths.put(ColumnPath.get(columnDescriptor.getPath()), columnDescriptor);
+        }
+        if (options.isUseColumnIndex() && parquetPredicate instanceof TupleDomainParquetPredicate) {
```

Comment:
The fact that we have to do the `instanceof` check and then downcast tells me there's something odd about the abstractions and structure of this code. Let me dig further to see how we can clean that up.

Reply:
Let me know if you find a better way.

Reply:
One option is to add a toParquetFilter() method to the Predicate interface, so this class doesn't need to know about the specific type of the Predicate instance. The method would return an `Optional<FilterPredicate>`